linker/shared/src/main/scala/org/scalajs/linker/backend/OutputWriter.scala

```diff
@@ -61,38 +69,52 @@ private[backend] abstract class OutputWriter(output: OutputDirectory,
     }
   }
 
-  private def writeModule(moduleID: ModuleID)(
+  private def writeModule(moduleID: ModuleID, existingFiles: Set[String])(
       implicit ec: ExecutionContext): Future[Report.Module] = {
     val jsFileName = OutputPatternsImpl.jsFile(config.outputPatterns, moduleID.id)
 
     if (config.sourceMap) {
       val sourceMapFileName = OutputPatternsImpl.sourceMapFile(config.outputPatterns, moduleID.id)
-
-      val codeWriter = new ByteArrayWriter
-      val smWriter = new ByteArrayWriter
-
-      writeModule(moduleID, codeWriter, smWriter)
-
-      val code = codeWriter.toByteBuffer()
-      val sourceMap = smWriter.toByteBuffer()
-
-      for {
-        _ <- outputImpl.writeFull(jsFileName, code)
-        _ <- outputImpl.writeFull(sourceMapFileName, sourceMap)
-      } yield {
-        new ReportImpl.ModuleImpl(moduleID.id, jsFileName, Some(sourceMapFileName), moduleKind)
+      val prepared = prepareModule(moduleID)
+      val report = new ReportImpl.ModuleImpl(moduleID.id, jsFileName, Some(sourceMapFileName), moduleKind)
+
+      if (!didModuleChange(moduleID, prepared) &&
+          existingFiles.contains(jsFileName) && existingFiles.contains(sourceMapFileName)) {
+        Future.successful(report)
+      } else {
+        val codeWriter = new ByteArrayWriter
+        val smWriter = new ByteArrayWriter
+
+        writeModule(moduleID, prepared, codeWriter, smWriter)
+
+        val code = codeWriter.toByteBuffer()
+        val sourceMap = smWriter.toByteBuffer()
+
+        for {
+          _ <- outputImpl.writeFull(jsFileName, code)
+          _ <- outputImpl.writeFull(sourceMapFileName, sourceMap)
```

Comment:
These will still read the target file first. Maybe it makes sense to bypass that (add a param) in incremental runs?

Reply:
No, I think we should keep it. On the first run in a new sbt session, you still want to avoid writing modules that were not changed since your last session.

Reply:
> On the first run in a new sbt session, you still want to avoid writing modules that were not changed since your last session.

Yes, of course. But in subsequent runs we wouldn't need to. And IIUC that's the case we'd want to optimizer performance for?

Reply:
Ah yes I see. I misunderstood. I can do that, indeed.